From 7190cc986337e43efbcba652435dc8ae03df43e0 Mon Sep 17 00:00:00 2001
From: t0yv0
Date: Sun, 4 Apr 2021 00:01:44 -0400
Subject: [PATCH] WIP on top of 27.1

---
 lisp/progmodes/compile.el | 24 +++++++++++++++++++++++-
 1 file changed, 23 insertions(+), 1 deletion(-)

diff --git a/lisp/progmodes/compile.el b/lisp/progmodes/compile.el
index 455f181f50..dbc4789ce0 100644
--- a/lisp/progmodes/compile.el
+++ b/lisp/progmodes/compile.el
@@ -1463,7 +1463,7 @@ to `compilation-error-regexp-alist' if RULES is nil."
         (error "HYPERLINK should be an integer: %s" (nth 5 item)))

       (goto-char start)
-      (while (re-search-forward pat end t)
+      (while (compilation--re-search-forward-limited pat end 1024)
         (when (setq props (compilation-error-properties
                            file line end-line col end-col (or type 2) fmt))

@@ -1524,6 +1524,28 @@ to `compilation-error-regexp-alist' if RULES is nil."
              (match-beginning mn) (match-end mn)
              'font-lock-face (cadr props))))))))

+(defun compilation--re-search-forward-limited (regexp bound n)
+  "Like 're-search-forward limited to the first N chars per line.
+This avoids Emacs performance degradation on scanning excessively
+long lines of text.  It is reasonable when scanning for compiler
+warnings to expect to find them early on each line.  REGEXP and
+BOUND are as in 're-search-forward."
+  (let ((inhibit-field-text-motion t)
+        (found nil)
+        (orig-point (point)))
+    (while (and (null found)
+                (< (point) bound)
+                (not (eobp)))
+      (setq found (re-search-forward regexp
+                                     (max (point)
+                                          (min bound (+ (point-at-bol) n)))
+                                     t))
+      (when (null found)
+        (forward-line 1)))
+    (when (null found)
+      (goto-char orig-point))
+    found))
+
 (defvar compilation--parsed -1)
 (make-variable-buffer-local 'compilation--parsed)

--
2.28.0
